woocommerce/woocommerce
# Add background HPOS cleanup process

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:




This PR adds a new tool to to WC > Status > Tools, that can be used to remove legacy data from orders. It triggers a background process (similar to the sync one) that removes data from the postmeta table for all orders. The new functionality is only available when HPOS is active and sync is disabled.

Closes #41915 & #41916.



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Build both the legacy assets and WC to make sure the JS and CSS changes are correct, by running `pnpm --filter='@woocommerce/plugin-woocommerce' build`.
1. Create a few orders in WC > Orders or use a tool such as Smooth Generator.
1. If necessary, run `wp wc cot sync` to sync orders.
1. Go to WC > Settings > Advanced > Features and make sure the HPOS datastore is selected and sync is disabled.
1. Go to WC > Orders and take note of a few order IDs.
1. Confirm the number of legacy order records on your site by running `wp post list --post_type=shop_order --format=count`. Take note of this number.
1. Confirm that your legacy records contain metadata by running `wp post meta list <order_id>` for a few of the order IDs from step 6.
1. Go to WC > Status > Tools and locate the "Clean up order data from legacy tables" tool.
   <img width="1353" alt="Screenshot 2024-01-26 at 17 23 12" src="https://github.com/woocommerce/woocommerce/assets/184724/9e8aeda5-22b8-4573-8706-0f2cca9ce7ca">
1. Click "Clear data" to start the process. The tool will now change to the following and kickstart a background process:
   <img width="1293" alt="Screenshot 2024-01-26 at 18 23 55" src="https://github.com/woocommerce/woocommerce/assets/184724/c4d8fa43-9d2b-4a19-b00f-586da2463ad8">
1. Feel free to test the "Cancel" functionality and restart the process as needed.
1. While the background process is running, you can run `wp post list --post_type=shop_order --format=count` to confirm that the number of orders from step 6 is decreasing. This should eventually reach 0.
1. Run `wp post list --post_type=shop_order_placehold --format=count` to confirm that all of the legacy records have been converted to placeholders.
1. Confirm that your legacy records now contain no metadata by running `wp post meta list <order_id>` with some of the order IDs from step 5.
1. Visit WC > Status > Tools again and confirm that the tool is disabled and that the button previously used to start the process now has a "No orders in need of cleanup" message:
   <img width="1342" alt="Screenshot 2024-01-26 at 17 30 11" src="https://github.com/woocommerce/woocommerce/assets/184724/15781576-384d-45cd-bb89-1b37d9bc442a">




### Changelog entry




<details>

#### Significance




#### Type




#### Message 

#### Comment 

</details>
